draft-ietf-quic-version-negotiation.md

```diff
@@ -225,42 +282,49 @@ received Version Negotiation packet, so clients MUST NOT add or remove reserved
-In particular, the server MUST ensure that there are no versions that it
-supports that are absent from the Received Negotiation Versions, and that the
-ordering matches the server's preference. If this validation fails, the server
-MUST close the connection with an error of type VERSION_NEGOTIATION_ERROR. This
+the version in the long header that carried the handshake version information.
+Similarly, clients MUST validate that the server's `Negotiated Version` matches
+the long header version. If an endpoint's validation fails, it MUST close the
+connection. If the connection was using QUIC version 1, it MUST be closed with
+a transport error of type `VERSION_NEGOTIATION_ERROR`.
+
+When a server parses the client's handshake version information, if the
+`Received Negotiation Version Count` is not zero, the server MUST validate that
+it could have sent the Version Negotation packet described by the client in
+response to a first flight of version `Previously Attempted Version`. In
+particular, the server MUST ensure that there are no versions that it supports
+that are absent from the Received Negotiation Versions, and that the ordering
+matches the server's preference. If this validation fails, the server MUST
+close the connection. If the connection was using QUIC version 1, it MUST be
+closed with a transport error of type `VERSION_NEGOTIATION_ERROR`. This
 mitigates an attacker's ability to forge Version Negotiation packets to force a
 version downgrade.
 
 If a server operator is progressively deploying a new QUIC version throughout
 its fleet, it MAY perform a two-step process where it first progressively adds
-support for the new version, but without enforcing its presence in Received
-Negotiation Versions. Once all servers have been upgraded, the second step is
-to start enforcing that the new version is present in Received Negotiation
-Versions. This opens connections to version downgrades during the upgrade
-window, since those could be due to clients communicating with both upgraded and
-non-upgraded servers.
+support for the new version, but without enforcing its presence in `Received
+Negotiation Versions`. Once all servers have been upgraded, the second step is
+to start enforcing that the new version is present in `Received Negotiation
+Versions`. This opens connections to version downgrades during the upgrade
+window, since those could be due to clients communicating with both upgraded
+and non-upgraded servers.
+
+If an endpoint receives its peer's Handshake Version Information and fails to
+parse it (for example, if it is too short), then the endpoint MUST close the
+connection. If the connection was using QUIC version 1, it MUST be closed with
+a transport error of type `TRANSPORT_PARAMETER_ERROR`.
 
 
 # Supported Versions
 
-The server's Supported Version list allows it to communicate the full list of
+The server's `Supported Version` list allows it to communicate the full list of
```

Comment:
I don't think this is that helpful and would remove it, but this PR doesn't change that, so we can deal with it separately.

Reply:
Yeah, please file a separate issue